lib/active_record_inherit_assoc.rb

```diff
@@ -29,6 +29,42 @@ def attribute_inheritance_hash
   prepend ActiveRecordInheritAssocPrepend
 end
 
+module ActiveRecordInheritPreloadAssocPrepend
+  def self.prepended(mod)
+    mod.class_eval do
+      alias_method_chain :associated_records_by_owner, :inherit
+    end
+  end
+
+  def filter_associated_records_with_inherit!(owner, associated_records)
+    return unless reflection.options[:inherit]
+
+    associated_records.select! do |record|
+      Array(reflection.options[:inherit]).all? do |obj|
+        record.send(obj) == owner.send(obj)
+      end
+    end
+  end
+
+  if ActiveRecord::VERSION::MAJOR >= 4 && ActiveRecord::VERSION::MINOR >= 1
```

Comment:
this instead of fancy &method ? `result.each { |r| filter_associated_records_with_inherit!(r) }`